File: blue-core/src/blue/orchestra/blueSynthBuilder/PresetGroup.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2000-2004 Steven Yi (stevenyi@gmail.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.orchestra.blueSynthBuilder;

import electric.xml.Element;
import electric.xml.Elements;
import java.util.ArrayList;
import java.util.Iterator;
import org.apache.commons.lang3.builder.EqualsBuilder;

/**
 * @author steven
 */
public class PresetGroup implements Comparable<PresetGroup> {

    private String presetGroupName = "Presets";

    private ArrayList<PresetGroup> subGroups = new ArrayList<>();

    private ArrayList<Preset> presets = new ArrayList<>();

    private String currentPresetUniqueId = null;

    private boolean currentPresetModified = false;

    public PresetGroup() {
    }

    public PresetGroup(PresetGroup presetGroup) {
        presetGroupName = presetGroup.presetGroupName;
        currentPresetUniqueId = presetGroup.currentPresetUniqueId;
        currentPresetModified = presetGroup.currentPresetModified;

        for (PresetGroup group : presetGroup.subGroups) {
            subGroups.add(new PresetGroup(group));
        }

        for (Preset preset : presetGroup.presets) {
            presets.add(new Preset(preset));
        }
    }

    public String getPresetGroupName() {
        return presetGroupName;
    }

    public void setPresetGroupName(String presetGroupName) {
        this.presetGroupName = presetGroupName;
    }

    public ArrayList<Preset> getPresets() {
        return presets;
    }

    public void setPresets(ArrayList<Preset> presets) {
        this.presets = presets;
    }

    public ArrayList<PresetGroup> getSubGroups() {
        return subGroups;
    }

    public void setSubGroups(ArrayList<PresetGroup> subGroups) {
        this.subGroups = subGroups;
    }

    /**
     * @return the currentPresetUniqueId
     */
    public String getCurrentPresetUniqueId() {
        return currentPresetUniqueId;
    }

    /**
     * @param currentPresetUniqueId the currentPresetUniqueId to set
     */
    public void setCurrentPresetUniqueId(String currentPresetUniqueId) {
        this.currentPresetUniqueId = currentPresetUniqueId;
    }

    /**
     * @return the currentPresetModified
     */
    public boolean isCurrentPresetModified() {
        return currentPresetModified;
    }

    /**
     * @param currentPresetModified the currentPresetModified to set
     */
    public void setCurrentPresetModified(boolean currentPresetModified) {
        this.currentPresetModified = currentPresetModified;
    }

    public static PresetGroup loadFromXML(Element data) {
        PresetGroup group = new PresetGroup();

        group.setPresetGroupName(data.getAttributeValue("name"));

        String val = data.getAttributeValue("currentPresetUniqueId");
        if (val != null && val.length() > 0) {
            group.setCurrentPresetUniqueId(val);
        }

        val = data.getAttributeValue("currentPresetModified");
        if (val != null && val.length() > 0) {
            group.setCurrentPresetModified(Boolean.valueOf(val).booleanValue());
        }

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            switch (node.getName()) {
                case "presetGroup":
                    PresetGroup pGroup = PresetGroup.loadFromXML(node);
                    group.getSubGroups().add(pGroup);
                    break;
                case "preset":
                    Preset preset = Preset.loadFromXML(node);
                    group.getPresets().add(preset);
                    break;
            }
        }

        return group;
    }

    public Element saveAsXML() {
        Element retVal = new Element("presetGroup");
        retVal.setAttribute("name", getPresetGroupName());

        if (currentPresetUniqueId != null) {
            retVal.setAttribute("currentPresetUniqueId", currentPresetUniqueId);
            retVal.setAttribute("currentPresetModified", Boolean.toString(currentPresetModified));
        }

        for (Iterator<PresetGroup> iter = subGroups.iterator(); iter.hasNext();) {
            PresetGroup subGroup = iter.next();

            retVal.addElement(subGroup.saveAsXML());
        }

        for (Iterator<Preset> iter = presets.iterator(); iter.hasNext();) {
            Preset preset = iter.next();

            retVal.addElement(preset.saveAsXML());
        }

        return retVal;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Comparable#compareTo(java.lang.Object)
     */
    @Override
    public int compareTo(PresetGroup b) {
        return this.getPresetGroupName().compareTo(b.getPresetGroupName());
    }

    /**
     * @param preset
     */
    public void addPreset(Preset preset) {
        presets.add(preset);
        // Collections.sort(presets);
    }

    /**
     * @param presetGroup
     */
    public void addPresetGroup(PresetGroup presetGroup) {
        subGroups.add(presetGroup);
        // Collections.sort(subGroups);
    }

    /**
     * @param preset
     */
    public boolean removePreset(Preset preset) {
        if (presets.contains(preset)) {
            presets.remove(preset);
            return true;
        }

        for (Iterator iter = subGroups.iterator(); iter.hasNext();) {
            PresetGroup tempGroup = (PresetGroup) iter.next();
            if (tempGroup.removePreset(preset)) {
                return true;
            }

        }

        return false;
    }

    /**
     * @param presetGroup
     */
    public boolean removePresetGroup(PresetGroup presetGroup) {
        if (subGroups.contains(presetGroup)) {
            subGroups.remove(presetGroup);
            return true;
        }

        for (Iterator<PresetGroup> iter = subGroups.iterator(); iter.hasNext();) {
            PresetGroup tempGroup = iter.next();

            if (tempGroup.removePresetGroup(presetGroup)) {
                return true;
            }
        }

        return false;
    }

    public Preset findPresetByUniqueId(String uniqueId) {

        if (uniqueId == null) {
            return null;
        }

        for (PresetGroup presetGroup : subGroups) {
            Preset preset = presetGroup.findPresetByUniqueId(uniqueId);
            if (preset != null) {
                return preset;
            }
        }

        for (Preset preset : presets) {
            if (uniqueId.equals(preset.getUniqueId())) {
                return preset;
            }
        }
        return null;
    }

    public String getPresetFullPathName(String uniqueId) {
        if (uniqueId == null) {
            return null;
        }

        for (PresetGroup presetGroup : subGroups) {
            Preset preset = presetGroup.findPresetByUniqueId(uniqueId);
            if (preset != null) {
                return presetGroup.getPresetGroupName() + " :: " + preset.getPresetName();
            }
        }

        for (Preset preset : presets) {
            if (uniqueId.equals(preset.getUniqueId())) {
                return preset.getPresetName();
            }
        }
        return null;
    }

    @Override
    public String toString() {
        return getPresetGroupName();
    }

    @Override
    public boolean equals(Object obj) {
        return EqualsBuilder.reflectionEquals(this, obj);
    }

}

File: packages/blue-data/src/instruments/blue-synth-builder/preset-group.ts
import { Element } from "../../serialization/xml-reader";
import type { BSBGraphicInterface } from "./bsb-graphic-interface";
import { Preset } from "./preset";

export class PresetGroup {
  presetGroupName = "Presets";
  subGroups: PresetGroup[] = [];
  presets: Preset[] = [];
  currentPresetUniqueId = "";
  currentPresetModified = false;

  getPresetGroupName(): string {
    return this.presetGroupName;
  }

  setPresetGroupName(name: string): void {
    this.presetGroupName = name;
  }

  getSubGroups(): PresetGroup[] {
    return [...this.subGroups];
  }

  getPresets(): Preset[] {
    return [...this.presets];
  }

  getCurrentPresetUniqueId(): string {
    return this.currentPresetUniqueId;
  }

  setCurrentPresetUniqueId(id: string): void {
    this.currentPresetUniqueId = id;
  }

  isCurrentPresetModified(): boolean {
    return this.currentPresetModified;
  }

  setCurrentPresetModified(modified: boolean): void {
    this.currentPresetModified = modified;
  }

  findPresetByUniqueId(uniqueId: string): Preset | null {
    for (const preset of this.presets) {
      if (preset.getUniqueId() === uniqueId) return preset;
    }
    for (const sub of this.subGroups) {
      const found = sub.findPresetByUniqueId(uniqueId);
      if (found) return found;
    }
    return null;
  }

  /**
   * Recursively synchronize every preset in this group tree with the
   * current BSB graphic interface.
   *
   * Mirrors Java `PresetsUtilities.synchronizePresets(PresetGroup, BSBGraphicInterface)`.
   */
  synchronizePresets(graphicInterface: BSBGraphicInterface): void {
    for (const subGroup of this.subGroups) {
      subGroup.synchronizePresets(graphicInterface);
    }
    for (const preset of this.presets) {
      preset.synchronizeWithInterface(graphicInterface);
    }
  }

  saveAsXML(): Element {
    const elem = new Element("presetGroup");
    elem.setAttribute("name", this.presetGroupName);
    if (this.currentPresetUniqueId) {
      elem.setAttribute("currentPresetUniqueId", this.currentPresetUniqueId);
    }
    elem.setAttribute("currentPresetModified", this.currentPresetModified.toString());
    for (const preset of this.presets) {
      elem.addElement(preset.saveAsXML());
    }
    for (const subGroup of this.subGroups) {
      elem.addElement(subGroup.saveAsXML());
    }
    return elem;
  }

  static loadFromXML(data: Element): PresetGroup {
    const group = new PresetGroup();
    group.presetGroupName = data.getAttribute("name") ?? "Presets";
    group.currentPresetUniqueId = data.getAttribute("currentPresetUniqueId") ?? "";
    const modified = data.getAttribute("currentPresetModified");
    group.currentPresetModified = modified === "true";

    const presetElems = data.getElements("preset");
    while (presetElems.hasMoreElements()) {
      group.presets.push(Preset.loadFromXML(presetElems.next()));
    }

    const subGroupElems = data.getElements("presetGroup");
    while (subGroupElems.hasMoreElements()) {
      group.subGroups.push(PresetGroup.loadFromXML(subGroupElems.next()));
    }

    return group;
  }

  private cloneForDuplicate(presetIdMap: Map<string, string>): PresetGroup {
    const copy = new PresetGroup();
    copy.presetGroupName = this.presetGroupName;
    copy.currentPresetUniqueId = this.currentPresetUniqueId;
    copy.currentPresetModified = this.currentPresetModified;
    copy.presets = this.presets.map((preset) => preset.deepCopy(presetIdMap));
    copy.subGroups = this.subGroups.map((group) => group.cloneForDuplicate(presetIdMap));
    return copy;
  }

  private rewriteCurrentPresetIds(presetIdMap: Map<string, string>): void {
    const nextCurrentId = presetIdMap.get(this.currentPresetUniqueId);
    if (nextCurrentId) {
      this.currentPresetUniqueId = nextCurrentId;
    }

    for (const group of this.subGroups) {
      group.rewriteCurrentPresetIds(presetIdMap);
    }
  }

  deepCopy(): PresetGroup {
    const presetIdMap = new Map<string, string>();
    const copy = this.cloneForDuplicate(presetIdMap);
    copy.rewriteCurrentPresetIds(presetIdMap);
    return copy;
  }
}
